src/snowflake/cli/api/entities/utils.py

```diff
@@ -46,35 +50,13 @@ def generic_sql_error_handler(
 ) -> NoReturn:
     # Potential refactor: If moving away from Python 3.8 and 3.9 to >= 3.10, use match ... case
     if err.errno == DOES_NOT_EXIST_OR_CANNOT_BE_PERFORMED:
-        raise ProgrammingError(
-            msg=dedent(
-                f"""\
-                Received error message '{err.msg}' while executing SQL statement.
-                '{role}' may not have access to warehouse '{warehouse}'.
-                Please grant usage privilege on warehouse to this role.
-                """
-            ),
-            errno=err.errno,
-        )
+        raise WarehouseDoesNotExistOrRoleUnauthorizedError(
+            warehouse=warehouse, role=role, msg=err.msg
+        ) from err
```

Comment:
We gotta do something about this generic error handler :D 

Let's start by renaming it, because it's not a generic _sql error_ handler; we are only invoking it when we catch _programming errors_ (there are other exception types that connectors can raise in response to a sql call). So it's really a programming error handler.

Secondly, it seems that all three handled error codes here (warehouse does not exist, no warehouse selected, and does not exist or unauthorized) are classified as "not our fault", since we are raising them as ClickExceptions. Is that correct? If yes, can we extract these checks into a util: "is it our fault"; and call it at the top of this programming error handler? That way, we can add more error types that are "not our fault" to that util. That helps translate the mental model of exception classes (def our fault, def theirs, we don't know) into code. It's not a perfect translation, but it's a start!

Lastly, same error code returned from connector can be caused in different contexts. e.g. 
`use warehouse non_existant_warehouse` will return the same error code as `use role non_existant_role`. So we need context passed in to this handler from where the programming error is caught into this handler to be able to raise the proper error type here. 
  
@sfc-gh-bdufour thoughts?

Reply:
Good thoughts, happy to discuss more, but I'd hold off on the suggested refactorings so that we can have a shot at including this at the very next opportunity (3.0.1 if possible).

Reply:
Sure. For the sake of getting this PR in then, I'd say at minimum do the renaming. I'd also like to also see the separation of the log based on the context (for the same error code), but since we are calling this handler only in specific places I'm ok with marking that as TODO and fixing it in a separate PR.

Reply:
Even the renaming doesn't buy us much. I'd fold that into the next work, since we'll pick it up right away.

Reply:
I added this comment thread to a new issue that I'll pick up right after, please feel free to add any other ideas: https://snowflakecomputing.atlassian.net/browse/SNOW-1722580